### PR TITLE
1st draft for review

### PR DESCRIPTION
**Pull request by ashtul**
@gkorland @itamarhaber @mnunberg @jeffreylovitz

### CODE REVIEW

File: docs/TopK_Commands.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# RedisBloom TopK Filter Command Documentation
```
Comment (itamarhaber) [inline docs/TopK_Commands.md]: Would be nice if you could add some time complexity information about each command, as well as what it actually does and returns.

Comment (ashtul) [reply docs/TopK_Commands.md]: Documentation updated as well as python unittests

File: docs/TopK_Commands.md [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+* **key**: Name of sketch where item is counted.
+* **item**: Item to be counted.
+
+### Complexity
+
+O(k + depth)
+
+### Return
+
+1 if item is in Top-K, otherwise 0.
```
Comment (jeffreylovitz) [inline docs/TopK_Commands.md]: Returns count, no?

File: docs/TopK_Commands.md [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+* **key**: Name of sketch where item is counted.
+
+### Complexity
+
+O(k)
+
+### Return
+
+k (or less) items in Top K list.
```
Comment (jeffreylovitz) [inline docs/TopK_Commands.md]: Why `(or less)`? Doesn't this function enumerate over the entire list?

File: src/rm_topk.c [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    RedisModule_ReplyWithError(ctx, x);                                                            \
+    return REDISMODULE_ERR;
+
+RedisModuleType *TopKType;
+
+static int GetTopKKey(RedisModuleCtx *ctx, RedisModuleString *keyName, TopK **topk, int mode) {
+    // All using this function should call RedisModule_AutoMemory to prevent memory leak
+    RedisModuleKey *key = RedisModule_OpenKey(ctx, keyName, mode);
+    if (RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
+        INNER_ERROR("TopK: key does not exist");
```
Comment (jeffreylovitz) [inline src/rm_topk.c]: Must call `RedisModule_CloseKey` before returning with an error.

File: src/rm_topk.c [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    } else if (RedisModule_ModuleTypeGetType(key) != TopKType) {
+        INNER_ERROR(REDISMODULE_ERRORMSG_WRONGTYPE);
```
Comment (jeffreylovitz) [inline src/rm_topk.c]: Must call `RedisModule_CloseKey` before returning with an error.

File: docs/TopK_Commands.md [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+Paper and additional information can be found [here](https://www.usenix.org/conference/atc18/presentation/gong).
+
+***
+
+## TOPK.RESERVE
+
+Initializes a TopK with specified parameters.
+
+```sql
+BF.RESERVE key topk width depth decay
```
Comment (jeffreylovitz) [inline docs/TopK_Commands.md]: `TOPK.RESERVE`

File: docs/TopK_Commands.md [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+### Complexity
+
+O(1)
+
+### Return
+
+Information.
+
+#### Reserve example
```
Comment (jeffreylovitz) [inline docs/TopK_Commands.md]: Why are all of these named "Reserve example"? I think that wording would only apply to TOPK.RESERVE - just having a generic "Example" would be fine. (For all commands in this documentation.)

File: docs/TopK_Commands.md [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+```sql
+TOPK.ADD test foo bar 42
+```
+
+***
+
+## TOPK.QUERY
+
+Checks whether an item is one of Top-K items.
```
Comment (jeffreylovitz) [inline docs/TopK_Commands.md]: I think the documentation should be more clear about which commands accept N arguments and return N-length arrays.

File: src/topk.c [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    assert(item);
+    assert(itemlen);
+
+    Bucket *runner = NULL;
+    uint32_t fp = TOPK_HASH(item, itemlen, GA);    
+    // TODO: The optimization of >heapMin should be revisited for performance
+    counter_t heapMin = checkExistInHeap(topk, item, itemlen) ? topk->heap->count : 0;
+    counter_t res = 0;
+
+    for(int i = 0; i < topk->depth; ++i) {
```
Comment (jeffreylovitz) [inline src/topk.c]: `uint32_t i`

File: src/test_topk.c [hunk 10]
```diff
@@ -0,0 +1,152 @@
+#include <stdio.h> // printf
```
Comment (jeffreylovitz) [inline src/test_topk.c]: Doesn't this file belong in the `tests` folder?

File: src/test_cms.c [hunk 11]
```diff
@@ -0,0 +1,101 @@
+
```
Comment (jeffreylovitz) [inline src/test_cms.c]: Doesn't this file belong in the `tests` folder?

File: src/rm_topk.c [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+        RedisModule_CloseKey(key);
+        return RedisModule_ReplyWithError(ctx, "TopK: key already exists");
+    }
+
+    TopK *topk = NULL;
+    if (createTopK(ctx, argv, argc, &topk) != REDISMODULE_OK)
+        return REDISMODULE_OK;
+
+    if (RedisModule_ModuleTypeSetValue(key, TopKType, topk) == REDISMODULE_ERR) {
+        return REDISMODULE_OK;
```
Comment (jeffreylovitz) [inline src/rm_topk.c]: Returning with unclosed key.

File: src/rm_topk.c [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+
+    RedisModuleKey *key = RedisModule_OpenKey(ctx, argv[1], REDISMODULE_READ | REDISMODULE_WRITE);
+    if (RedisModule_KeyType(key) != REDISMODULE_KEYTYPE_EMPTY) {
+        RedisModule_CloseKey(key);
+        return RedisModule_ReplyWithError(ctx, "TopK: key already exists");
+    }
+
+    TopK *topk = NULL;
+    if (createTopK(ctx, argv, argc, &topk) != REDISMODULE_OK)
+        return REDISMODULE_OK;
```
Comment (jeffreylovitz) [inline src/rm_topk.c]: Returning with unclosed key.

File: src/rm_topk.c [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+    }
+    if ((RedisModule_StringToLongLong(argv[4], &depth) != REDISMODULE_OK) || depth < 1) {
+        INNER_ERROR("TopK: invalid depth");
+    }    
+    if ((RedisModule_StringToDouble(argv[5], &decay) != REDISMODULE_OK) ||
+                                    (decay <= 0 || decay >= 1)) {
+        INNER_ERROR("TopK: invalid decay value. must be '> 1' & '< 0'");
+    }
+
+    *topk = TopK_Create(k, width, depth, decay);
```
Comment (jeffreylovitz) [inline src/rm_topk.c]: The level of strictness you want to maintain about this is up to you, but your long longs are implicitly getting demoted to unsigned ints here. Which is likely fine, but good to be aware of!

File: src/rm_topk.c [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+    }
+    if ((RedisModule_StringToLongLong(argv[3], &width) != REDISMODULE_OK) || width < 1) {
+        INNER_ERROR("TopK: invalid width");
+    }
+    if ((RedisModule_StringToLongLong(argv[4], &depth) != REDISMODULE_OK) || depth < 1) {
+        INNER_ERROR("TopK: invalid depth");
+    }    
+    if ((RedisModule_StringToDouble(argv[5], &decay) != REDISMODULE_OK) ||
+                                    (decay <= 0 || decay >= 1)) {
+        INNER_ERROR("TopK: invalid decay value. must be '> 1' & '< 0'");
```
Comment (jeffreylovitz) [inline src/rm_topk.c]: Inequality reversed in error message.

File: src/rm_topk.c [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+    } else if (RedisModule_ModuleTypeGetType(key) != TopKType) {
+        INNER_ERROR(REDISMODULE_ERRORMSG_WRONGTYPE);
+    }
+    *topk = RedisModule_ModuleTypeGetValue(key);
+    RedisModule_CloseKey(key); //TODO: check if this works
+    return REDISMODULE_OK;
+}
+
+
+static int createTopK(RedisModuleCtx *ctx, RedisModuleString **argv, int argc, TopK **topk) {
```
Comment (jeffreylovitz) [inline src/rm_topk.c]: `argc` not needed here (arity is validated by caller already)

File: src/rm_topk.c [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+/**************** Module functions *********************************/
+
+void TopKRdbSave(RedisModuleIO *io, void *obj) {
+    TopK *topk = obj;
+    RedisModule_SaveUnsigned(io, topk->k);
+    RedisModule_SaveUnsigned(io, topk->width);
+    RedisModule_SaveUnsigned(io, topk->depth);
+    RedisModule_SaveDouble(io, topk->decay);
+    RedisModule_SaveStringBuffer(io, (const char *)topk->data,
+                                 topk->width * topk->depth * sizeof(Bucket));
```
Comment (jeffreylovitz) [inline src/rm_topk.c]: There seems to be a potential memory error here. To illustrate in the simplest case:
`$ valgrind redis-server --loadmodule ./rebloom.so`
`$ redis-cli topk.reserve topk 20 50 5 0.9`
`$ redis-cli save`

Logs the output:
```
==15047== Conditional jump or move depends on uninitialised value(s)
==15047==    at 0x14790C: lzf_compress (lzf_c.c:153)
==15047==    by 0x15F505: rdbSaveLzfStringObject.part.2 (rdb.c:349)
==15047==    by 0x15F90C: rdbSaveLzfStringObject (rdb.c:346)
==15047==    by 0x15F90C: rdbSaveRawString (rdb.c:421)
==15047==    by 0x1B1873: RM_SaveStringBuffer (module.c:3153)
==15047==    by 0x83E78E3: TopKRdbSave (rm_topk.c:193)
==15047==    by 0x160AAD: rdbSaveObject (rdb.c:978)
==15047==    by 0x1614FD: rdbSaveKeyValuePair (rdb.c:1041)
==15047==    by 0x161BBC: rdbSaveRio (rdb.c:1145)
==15047==    by 0x16210A: rdbSave (rdb.c:1244)
==15047==    by 0x1650CE: saveCommand (rdb.c:2396)
==15047==    by 0x13FC36: call (server.c:2437)
==15047==    by 0x14052D: processCommand (server.c:2729)
```

Running the `topk.py` tests under Valgrind report further errors running count queries after a DEBUG RELOAD, but the root cause is probably here.

File: docs/TopK_Commands.md (hunk 3 above)
Comment (ashtul) [reply docs/TopK_Commands.md]: If LIST is called before the list is full (must be early on as, in the beginning, all items will go in).
I didn't like the empty cells so I'm only returning existing items.
It is a matter of a decision and I am fine with it one way or another.

File: src/rm_topk.c (hunk 14 above)
Comment (ashtul) [reply src/rm_topk.c]: I'm using long long b/c it is converted from redisStrings and that is my only option for an unsigned variable.

File: src/rm_topk.c (hunk 4 above)
Comment (ashtul) [reply src/rm_topk.c]: I've used RedisModule_AutoMemory. Will change to manual closing.

File: src/rm_topk.c (hunk 17 above)
Comment (ashtul) [reply src/rm_topk.c]: This is a 'feature' of the module itself.
Opened an issue, but shouldn't effect this PR.

File: src/test_topk.c (hunk 10 above)
Comment (ashtul) [reply src/test_topk.c]: Moved

File: docs/TopK_Commands.md (hunk 8 above)
Comment (ashtul) [reply docs/TopK_Commands.md]: How is it now?

**Comment by ashtul** [comment]
@jeffreylovitz @mnunberg @gkorland 
Work done on all comments.

Valgrind warning isn't caused by this PR. Will be dealt with separately.
whether TOPK.LIST returns shirked list is a matter of decision. It is only important at the beginning when some heap buckets are empty. Soon enough it is all full which eliminates the need for such a measure.

**Comment by ashtul** [comment]
Last few commits add 'make coverage' option

File: src/rm_topk.c (hunk 14 above)
Comment (jeffreylovitz) [reply src/rm_topk.c]: Yeah, the module API is unfortunately a bit lacking with regard to comprehensive type handling. The safer alternative would be checking like:
```
    if ((RedisModule_StringToLongLong(argv[3], &width) != REDISMODULE_OK) || width < 1 || width > UINT_MAX) {
        INNER_ERROR("TopK: invalid width");
```
But that's definitely a standard of safety we don't always hold ourselves to.

No changes needed, I just wanted to highlight this area!

File: src/rm_topk.c (hunk 17 above)
Comment (jeffreylovitz) [reply src/rm_topk.c]: :+1: Noted.

This issue should be investigated separately soon, though.

File: docs/TopK_Commands.md (hunk 3 above)
Comment (jeffreylovitz) [reply docs/TopK_Commands.md]: :+1: That makes sense, no need to make changes.

File: docs/TopK_Commands.md (hunk 8 above)
Comment (jeffreylovitz) [reply docs/TopK_Commands.md]: Much better, thanks!

**Pull request merged**